File: Assignment4/12/WindowsFormsApp1_Clock/Form1.cs
using System;
using System.Drawing;
using System.Windows.Forms;
using System.Timers;

namespace WindowsFormsAlarmApp
{
    public partial class Form1 : Form
    {
        private System.Timers.Timer timer;
        private DateTime targetTime;
        private Random random = new Random();
        private bool isAlarmSet = false;
        private TextBox textBoxTime;
        private Button buttonStart;

        public Form1()
        {
            // Manually set up the form
            this.Text = "Alarm App";
            this.Size = new Size(300, 200);

            // Create and configure controls
            textBoxTime = new TextBox
            {
                Name = "textBoxTime",
                Location = new Point(50, 50),
                Size = new Size(100, 20)
            };

            buttonStart = new Button
            {
                Name = "buttonStart",
                Text = "Start",
                Location = new Point(50, 80),
                Size = new Size(100, 30)
            };

            // Wire up the button click event
            buttonStart.Click += buttonStart_Click;

            // Add controls to the form
            this.Controls.Add(new Label
            {
                Text = "Enter time (HH:MM:SS):",
                Location = new Point(50, 20),
                AutoSize = true
            });
            this.Controls.Add(textBoxTime);
            this.Controls.Add(buttonStart);
        }

        private void buttonStart_Click(object sender, EventArgs e)
        {
            if (!isAlarmSet)
            {
                // Validate and parse input time
                if (DateTime.TryParse(textBoxTime.Text, out targetTime))
                {
                    targetTime = DateTime.Today.Add(targetTime.TimeOfDay);
                    if (targetTime < DateTime.Now) // If time is in the past, set for tomorrow
                        targetTime = targetTime.AddDays(1);

                    // Disable controls to prevent further input
                    textBoxTime.Enabled = false;
                    buttonStart.Text = "Alarm Set";
                    isAlarmSet = true;

                    // Start timer to check time and change color
                    timer = new System.Timers.Timer(1000);
                    timer.Elapsed += Timer_Tick;
                    timer.AutoReset = true;
                    timer.Start();
                }
                else
                {
                    MessageBox.Show("Invalid time format. Please use HH:MM:SS.", "Error");
                }
            }
        }

        private void Timer_Tick(object sender, ElapsedEventArgs e)
        {
            // Run UI updates on the UI thread
            if (this.InvokeRequired)
            {
                this.Invoke(new MethodInvoker(() =>
                {
                    this.BackColor = Color.FromArgb(random.Next(256), random.Next(256), random.Next(256));

                    // Check if current time matches target time
                    if (DateTime.Now.Hour == targetTime.Hour &&
                        DateTime.Now.Minute == targetTime.Minute &&
                        DateTime.Now.Second == targetTime.Second)
                    {
                        timer.Stop();
                        MessageBox.Show("Alarm! Time's up!", "Alarm");
                        this.BackColor = SystemColors.Control; // Reset color
                        textBoxTime.Enabled = true;
                        buttonStart.Text = "Start";
                        isAlarmSet = false;
                    }
                }));
            }
        }
    }
}

File: Assignment4/12/WindowsFormsApp_Clock1/Form1.vb
Imports System
Imports System.Drawing
Imports System.Windows.Forms
Imports System.Timers

Public Class Form1
    Private timer As System.Timers.Timer
    Private targetTime As DateTime
    Private random As New Random()
    Private isAlarmSet As Boolean = False
    Private Const PLACEHOLDER_TEXT As String = "Enter time (HH:MM:SS, 24-hour format)"

    Private Sub TextBox1_Enter(sender As Object, e As EventArgs)
        If TextBox1.Text = PLACEHOLDER_TEXT Then
            TextBox1.Text = ""
            TextBox1.ForeColor = SystemColors.WindowText ' Restore normal text color
        End If
    End Sub

    Private Sub TextBox1_Leave(sender As Object, e As EventArgs)
        If String.IsNullOrWhiteSpace(TextBox1.Text) Then
            TextBox1.Text = PLACEHOLDER_TEXT
            TextBox1.ForeColor = SystemColors.GrayText ' Gray text for placeholder
        End If
    End Sub

    Private Sub Button1_Click(sender As Object, e As EventArgs) Handles Button1.Click
        If Not isAlarmSet Then
            ' Check if the text is the placeholder before parsing
            If TextBox1.Text = PLACEHOLDER_TEXT Then
                MessageBox.Show("Please enter a valid time.", "Error")
                Exit Sub
            End If

            ' Validate and parse input time
            If DateTime.TryParse(TextBox1.Text, targetTime) Then
                targetTime = DateTime.Today.Add(targetTime.TimeOfDay)
                If targetTime < DateTime.Now Then ' If time is in the past, set for tomorrow
                    targetTime = targetTime.AddDays(1)
                End If

                ' Disable controls to prevent further input
                TextBox1.Enabled = False
                Button1.Text = "Alarm Set"
                isAlarmSet = True

                ' Start timer to check time and change color
                timer = New System.Timers.Timer(1000)
                AddHandler timer.Elapsed, AddressOf Timer_Tick
                timer.AutoReset = True
                timer.Start()
            Else
                MessageBox.Show("Invalid time format. Please use HH:MM:SS.", "Error")
            End If
        End If
    End Sub

    Private Sub Timer_Tick(sender As Object, e As ElapsedEventArgs)
        If Me.InvokeRequired Then
            Me.Invoke(New MethodInvoker(Sub()
                                            Me.BackColor = Color.FromArgb(random.Next(256), random.Next(256), random.Next(256))

                                            ' Check if current time matches target time
                                            If DateTime.Now.Hour = targetTime.Hour AndAlso
                                               DateTime.Now.Minute = targetTime.Minute AndAlso
                                               DateTime.Now.Second = targetTime.Second Then
                                                timer.Stop()
                                                MessageBox.Show("Alarm! Time's up!", "Alarm")
                                                Me.BackColor = SystemColors.Control ' Reset color
                                                TextBox1.Enabled = True
                                                Button1.Text = "Start"
                                                isAlarmSet = False
                                                TextBox1.Text = PLACEHOLDER_TEXT ' Restore placeholder
                                                TextBox1.ForeColor = SystemColors.GrayText ' Restore placeholder color
                                            End If
                                        End Sub))
        End If
    End Sub

    Private Sub Form1_Load(sender As Object, e As EventArgs) Handles MyBase.Load

    End Sub
End Class